btree/btree.c

```diff
@@ -0,0 +1,13031 @@
+	 */
+	M0_PRE(oi->i_extra_node != NULL);
+	bnode_lock(lev->l_node);
+	bnode_lock(oi->i_extra_node);
+
+	bnode_set_level(oi->i_extra_node, curr_max_level);
+
+	bnode_move(lev->l_node, oi->i_extra_node, D_RIGHT, NR_MAX);
+	M0_ASSERT(bnode_count_rec(lev->l_node) == 0);
+
+	bnode_set_level(lev->l_node, curr_max_level + 1);
+
+	/* 2) add new 2 records at root node. */
+
+	/* Add first rec at root */
+	node_slot = (struct slot) {
+		.s_node = lev->l_node,
+		.s_idx  = 0
+	};
+	node_slot.s_rec = bop->bo_rec;
+
+	/* M0_ASSERT(bnode_isfit(&node_slot)) */
+	bnode_make(&node_slot);
+	REC_INIT(&node_slot.s_rec, &p_key, &ksize, &p_val, &vsize);
+	bnode_rec(&node_slot);
+	COPY_RECORD(&node_slot.s_rec, &bop->bo_rec);
+	/* if we need to update vec_count for node, update here */
+
+	bnode_done(&node_slot, true);
+
+	/* Add second rec at root */
+
+	/**
+	 * For second record, value = segaddr(i_extra_node).
+	 * Note that, last key is not considered as valid key for internal node.
+	 * Therefore, currently, key is not set as NULL explicitly.
+	 * In future, depending on requirement, key, key size, value size might
+	 * need to be set/store explicitly.
+	 */
+	bop->bo_rec.r_val.ov_buf[0] = &(oi->i_extra_node->n_addr);
+	node_slot.s_idx  = 1;
+	node_slot.s_rec = bop->bo_rec;
+	bnode_make(&node_slot);
+	REC_INIT(&node_slot.s_rec, &p_key, &ksize, &p_val, &vsize);
+	bnode_rec(&node_slot);
+	COPY_VALUE(&node_slot.s_rec, &bop->bo_rec);
+	/* if we need to update vec_count for root slot, update at this place */
+
+	bnode_done(&node_slot, true);
+	bnode_seq_cnt_update(lev->l_node);
+	bnode_fix(lev->l_node);
+	/**
+	 * Note : not capturing l_node as it must have already been captured in
+	 * btree_put_makespace_phase().
+	 */
+	btree_node_capture_enlist(oi, oi->i_extra_node, 0);
+
+	/* Increase height by one */
+	tree->t_height++;
+	bop->bo_arbor->t_height = tree->t_height;
+	if (tree->t_height > MAX_TREE_HEIGHT) {
+		M0_LOG(M0_ERROR,
+		       "Tree height increased beyond MAX_TREE_HEIGHT");
+		M0_ASSERT(0);
+	}
+	/* Capture this change in transaction */
+
+	/* TBD : This check needs to be removed when debugging is done. */
+	M0_ASSERT_EX(bnode_expensive_invariant(lev->l_node));
+	M0_ASSERT_EX(bnode_expensive_invariant(oi->i_extra_node));
+	bnode_unlock(lev->l_node);
+	bnode_unlock(oi->i_extra_node);
+
+	return P_CAPTURE;
+}
+
+/**
+ * This function is called when there is overflow and splitting needs to be
+ * done. It will move some records from right node(l_node) to left node(l_alloc)
+ * and find the appropriate slot for given record. It will store the node and
+ * index (where we need to insert given record) in tgt slot as a result.
+ *
+ * @param l_alloc It is the newly allocated node, where we want to move record.
+ * @param l_node It is the current node, from where we want to move record.
+ * @param rec It is the given record for which we want to find slot
+ * @param tgt result of record find will get stored in tgt slot
+ */
```

Comment:
Minor: In the above comment need to replace l_node with current_node (right) and l_alloc with allocated_node (left)

Reply:
Done.

Reply:
Done